token-client/src/main/java/com/sap/cloud/security/xsuaa/client/DefaultOAuth2TokenService.java

```diff
@@ -0,0 +1,201 @@
+						"Error retrieving JWT token. Received status code %s. Call to XSUAA was not successful. Client credentials invalid.",
+						responseStatusCode));
+			}
+			if (!responseStatusCode.is2xxSuccessful()) {
+				throw new OAuth2ServiceException(String.format(
+						"Error retrieving JWT token. Received status code %s. Call to XSUAA was not successful.",
+						responseStatusCode));
+			}
+			throw new OAuth2ServiceException(String.format(
+					"Error retrieving JWT token. Call to XSUAA was not successful: %s",
+					ex.getMessage()));
+		}
+
+		@SuppressWarnings("unchecked")
+		Map<String, String> accessTokenMap = responseEntity.getBody();
+
+		String accessToken = accessTokenMap.get(ACCESS_TOKEN);
+		long expiresIn = Long.parseLong(String.valueOf(accessTokenMap.get(EXPIRES_IN)));
+		String refreshToken = accessTokenMap.get(REFRESH_TOKEN);
+		return new OAuth2AccessToken(accessToken, expiresIn, refreshToken);
+	}
+
+	/**
+	 * Create a copy of the given map or an new empty map
+	 * 
+	 * @return a new Map that contains all entries of the optional map
+	 */
+	private static Map<String, String> copy(Map<String, String> map) {
+		return map == null ? new HashMap<>() : new HashMap<>(map);
+	}
+
+	/**
+	 * Creates the set of HTTP headers with client-credentials basic authentication
+	 * header.
+	 *
+	 * @return the HTTP headers.
+	 */
+	private static HttpHeaders createHeadersWithAuthorization(ClientCredentials credentials) {
+		HttpHeaders headers = new HttpHeaders();
+		addAcceptHeader(headers);
+		addBasicAuthHeader(headers, credentials);
+		return headers;
+	}
+
+	/**
+	 * Creates the set of HTTP headers with Authorization Bearer header.
+	 *
+	 * @return the HTTP headers.
+	 */
+	private static HttpHeaders createHeadersWithAuthorization(String token) {
+		HttpHeaders headers = new HttpHeaders();
+		addAcceptHeader(headers);
+		addAuthorizationBearerHeader(headers, token);
+		return headers;
+	}
+
+	/** common utilities **/
+
+	/**
+	 * Adds the {@code  Accept: application/json} header to the set of headers.
+	 *
+	 * @param headers
+	 *            - the set of headers to add the header to.
+	 */
+	static void addAcceptHeader(HttpHeaders headers) {
+		headers.add(HttpHeaders.ACCEPT, MediaType.APPLICATION_JSON_VALUE);
+	}
+
+	/**
+	 * Adds the {@code  Authorization: Basic <credentials>} header to the set of
+	 * headers.
+	 *
+	 * @param headers
+	 *            - the set of headers to add the header to.
+	 * @param credentials
+	 *            - the client credentials used for authentication.
+	 */
+	static void addBasicAuthHeader(HttpHeaders headers, ClientCredentials credentials) {
+		final String BASIC_AUTH_HEADER_FORMAT = "Basic %s";
+		final String CREDENTIALS_FORMAT = "%s:%s";
+
+		String credentialsString = String
+				.format(CREDENTIALS_FORMAT, credentials.getId(), credentials.getSecret());
+		String base64Creds = Base64.getEncoder().encodeToString(credentialsString.getBytes(StandardCharsets.UTF_8));
```

Comment:
see diference https://www.pivotaltracker.com/n/projects/997278/stories/166970393

Reply:
thanks a lot for this hint. Its done with this commit:
https://github.com/SAP/cloud-security-xsuaa-integration/pull/124/commits/44bc95c7e5452af60373e0984f278f3d38d674ea